modules/040-node-manager/images/bashible-apiserver/pkg/template/steps_storage.go

```diff
@@ -17,48 +17,195 @@ limitations under the License.
 	if err != nil {
 		return nil, err
 	}
 
-	steps := map[string]string{}
+	if len(ng) > 0 {
+		userConfigurations, err := s.renderNodeGroupConfigurations(bundle, ng[0], templateContext)
+		if err != nil {
+			klog.Errorf("Render user NodeGroupConfigurations failed: %s", err)
+			return steps, nil
+		}
+
+		for k, v := range userConfigurations {
+			if _, ok := steps[k]; ok {
+				klog.Errorf("NodeGroupConfigurations conflicts with system script: %s", k)
+				continue
+			}
+			steps[k] = v
+		}
+	}
+
+	return steps, nil
+}
+
+func (s *StepsStorage) OnNodeConfigurationsChanged() chan struct{} {
+	return s.configurationsChanged
+}
+
+func (s *StepsStorage) runBufferedEmitter() chan struct{} {
+	for {
+		time.Sleep(500 * time.Millisecond)
+		s.emitter.Lock()
+		if s.emitter.count > 0 {
+			s.configurationsChanged <- struct{}{}
+			s.emitter.count = 0
+		}
+		s.emitter.Unlock()
+	}
+}
+
+func (s *StepsStorage) emitChanges() {
+	s.emitter.Lock()
+	s.emitter.count++
+	s.emitter.Unlock()
+}
+
+func (s *StepsStorage) subscribeOnCRD(ctx context.Context, ngConfigFactory dynamicinformer.DynamicSharedInformerFactory) {
+	if ngConfigFactory == nil {
+		return
+	}
+
+	go s.runBufferedEmitter()
+
+	// Launch the informer
+	ginformer := ngConfigFactory.ForResource(schema.GroupVersionResource{
+		Group:    "deckhouse.io",
+		Version:  "v1alpha1",
+		Resource: "nodegroupconfigurations",
+	})
+
+	informer := ginformer.Informer()
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			unst := obj.(*unstructured.Unstructured)
+			var ngc NodeGroupConfiguration
+			err := fromUnstructured(unst, &ngc)
+			if err != nil {
+				klog.Errorf("Convert from unstructured failed: %s", err)
+				return
+			}
+			s.AddNodeGroupConfiguration(&ngc)
+			s.emitChanges()
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			newUn := newObj.(*unstructured.Unstructured)
+			oldUn := oldObj.(*unstructured.Unstructured)
+
+			var newConf NodeGroupConfiguration
+			err := fromUnstructured(newUn, &newConf)
+			if err != nil {
+				klog.Errorf("Convert from unstructured failed: %s", err)
+				return
+			}
+
+			var oldConf NodeGroupConfiguration
+			err = fromUnstructured(oldUn, &oldConf)
+			if err != nil {
+				klog.Errorf("Convert from unstructured failed: %s", err)
+				return
+			}
+
+			if newConf.Spec.IsEqual(oldConf.Spec) {
+				return
+			}
+
+			s.RemoveNodeGroupConfiguration(&oldConf)
+			s.AddNodeGroupConfiguration(&newConf)
+			s.emitChanges()
+		},
+		DeleteFunc: func(obj interface{}) {
+			unst := obj.(*unstructured.Unstructured)
+			var ngc NodeGroupConfiguration
+			err := fromUnstructured(unst, &ngc)
+			if err != nil {
+				klog.Errorf("Convert from unstructured failed: %s", err)
+				return
+			}
+			s.RemoveNodeGroupConfiguration(&ngc)
+			s.emitChanges()
+		},
+	})
+
+	go informer.Run(ctx.Done())
+
+	// Wait for the first sync of the informer cache, should not take long
+	for !informer.HasSynced() {
```

Comment:
There is a function in client-go especially for this case, wich is also provides a circuit breaker:
```go
if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
    return nil, fmt.Errorf("unable to sync caches: %v", ctx.Err())
}
```